apiutil/apiutil.go

```diff
@@ -0,0 +1,155 @@
+// Package apiutil provides a set of functions for Bearer and other
+// authentication across our APIs and services
+package apiutil // import "github.com/teamwork/utils/apiutil"
+
+import (
+	"crypto/aes"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"encoding/hex"
+
+	"encoding/base64"
+
+	"github.com/andreburgaud/crypt2go/ecb"
+	"github.com/andreburgaud/crypt2go/padding"
+	"github.com/jmcvetta/randutil"
+	"github.com/pkg/errors"
+)
+
+// TokenData is used for generating and parsing Projects-style API tokens.
+type TokenData struct {
+	InstallationID int64 `json:"installationId"`
+	Shard          int64 `json:"shard"`
+	UserID         int64 `json:"userId"`
+	ObjectID       int64 `json:"objectId"`
+}
+
+// GenerateAPIAuthToken is ported from Projects (precise function is
+// cfcs/utility/security.cfc@generateSecureAPIToken). It generates a secure(ish)
+// permanent token that can be used to authenticate with Projects API.
+// These will be used as part of app login flow until a more robust OAuth2
+// or API key system is implemented and supported across all products.
+func GenerateAPIAuthToken(aesKey string, tokenData *TokenData) (string, error) {
+	if tokenData == nil {
+		return "", errors.New("nil pointer provided for tokenData")
+	}
```

Comment:
I wonder if it'd also be worth checking each of the fields, and returning an error if any are empty? We do the same on the decrypt side, but I don't see any safeguard against creating "invalid" tokens.

Reply:
We can I guess 😄